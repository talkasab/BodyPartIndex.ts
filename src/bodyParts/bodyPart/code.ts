import { ICode } from 'src/bodyParts/interfaces/ICode';

export class Code {

    /**
	 * The code constructor initialization.
	 * @param {ICode} data - The data.
	 */
    public constructor (
		private data: ICode
    ) {}

    /**
	 * Returns the code.
	 */
    public getCode (): string {
        return this.data.code;
    }

    /**
	 * Returns the system.
	 */
    public getSystem (): string {
        return this.data.system;
    }

}