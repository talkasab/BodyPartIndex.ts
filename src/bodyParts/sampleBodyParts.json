{
    "$schema": "MY_SCHEMA",
    "$version": "2.0.0",
    "bodyParts": [ 
		{
			"codes": [ { "code": "78320081", "system": "SNOMED" } ],
			"containedById": "RID901",
			"description": "hand",
			"partOfId": "RID902",
			"radlexId": "RID901"
		},
		{
			"codes": [ { "code": "78320082", "system": "SNOMED" } ],
			"containedById": "RID901",
			"description": "feet",
			"radlexId": "RID902"
		},
		{
			"codes": [ { "code": "78320083", "system": "SNOMED" } ],
			"containedById": "RID901",
			"description": "leg",
			"radlexId": "RID903",
			"synonyms": [ "feet" ]
		},
		{
			"containedById": "RID903",
			"description": "eye",
			"radlexId": "RID904",
			"synonyms": [ "aaa", "ccc" ],
			"partOfId": "RID902"
		},
		{
			"codes": [ { "code": "78320085", "system": "SNOMED" } ],
			"containedById": "RID904",
			"description": "finger",
			"partOfId": "RID901",
			"radlexId": "RID905"
		}
	]
}