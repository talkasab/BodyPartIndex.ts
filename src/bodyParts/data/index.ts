import { IBodyPart } from 'src/interfaces/bodyParts/IBodyPart';
import { IBodyPartsFile } from 'src/interfaces/bodyParts/IBodyPartsFile';

type AncestorsMap = { [key: string]: string };
type ChildrenMap = { [key: string]: string[] };
type BodyPartsMap = { [ key: string]: IBodyPart };

class BodyPartsData {

    public static version: string;
    public static bodyPartsMap: BodyPartsMap;
    public static containedAncestors: AncestorsMap;
    public static containedChildren: ChildrenMap;
    public static partOfAncestors: AncestorsMap;
    public static partOfChildren: ChildrenMap;

    /**
     * Initializes the BodyPartsData class.
     * @param {IBodyPartsFile} file - The file.
     */
    public static init (file: IBodyPartsFile): void {
        const containedAncestors: AncestorsMap = {};
        const containedChildren: ChildrenMap = {};
        const partOfAncestors: AncestorsMap = {};
        const partOfChildren: ChildrenMap = {};
        const bodyPartsMap: BodyPartsMap = {};

        file.bodyParts.forEach(item => {
            bodyPartsMap[item.radlexId] = item;
            this.initContained(containedAncestors, containedChildren, item);
            this.initPartOf(partOfAncestors, partOfChildren, item);
        });

        this.version = file.$version;
        this.bodyPartsMap = bodyPartsMap;
        this.containedAncestors = containedAncestors;
        this.containedChildren = containedChildren;
        this.partOfAncestors = partOfAncestors;
        this.partOfChildren = partOfChildren;
    }

    /**
	 * Initializes the contained hierarchy.
	 * @param {AncestorsMap} containedAncestors - The contained ancestors.
	 * @param {ChildrenMap} containedChildren - The contained children. 
	 * @param {IBodyPart} item - The item.
	 */
    private static initContained (containedAncestors: AncestorsMap, containedChildren: ChildrenMap, item: IBodyPart): void {
        containedAncestors[item.radlexId] = item.containedById;
			
        if (!containedChildren[item.containedById]) {
            containedChildren[item.containedById] = [];
        }

        if (item.containedById !== item.radlexId) {
            containedChildren[item.containedById].push(item.radlexId);
        }
    }

    /**
	 * Initializes the partOf hierarchy.
	 * @param {AncestorsMap} partOfAncestors - The partOf ancestors.
	 * @param {ChildrenMap} partOfChildren - The partOf children. 
	 * @param {IBodyPart} item - The item.
	 */
	 private static initPartOf (partOfAncestors: AncestorsMap, partOfChildren: ChildrenMap, item: IBodyPart): void {
        if (!item.partOfId) return;

        partOfAncestors[item.radlexId] = item.partOfId;
			
        if (!partOfChildren[item.partOfId]) {
            partOfChildren[item.partOfId] = [];
        }

        if (item.partOfId !== item.radlexId) {
            partOfChildren[item.partOfId].push(item.radlexId);
        }
    }
	
}

export {
    BodyPartsData
};