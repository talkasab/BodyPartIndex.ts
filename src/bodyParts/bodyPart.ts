import { IBodyPart } from 'src/interfaces/bodyParts/IBodyPart';
import { BodyPartsMap, BodyPartsUtils } from 'src/bodyParts/utils';
import { BodyPartsData } from 'src/bodyParts/data';

class BodyPart {

    /**
	 * Initializes the BodyPart.
	 * @param {IBodyPart} data - The data.
	 */
    public constructor (
		private data: IBodyPart
    ) { }

    /**
	 * Returns the data.
	 */
    public getData (): IBodyPart {
        return this.data;
    }

    /**
	 * Returns the immediate contained children.
	 */
    public getImmediateContainedChildren (): BodyPart[] {
        return this.getAllContainedChildren(1);
    }

    /**
	 * Returns the children.
	 * @param {number} maxLevel - The maxLevel (if specified will return children up to that level).
	 */
	 public getAllContainedChildren (maxLevel = -1): BodyPart[] {
        const immediateChildren = BodyPartsData.containedChildren[this.data.radlexId];
        const map: BodyPartsMap = {};
		
        if (!immediateChildren) return [];

        const traverse = (children: string[], current: number): void => {
            children.forEach(child => {
                map[child] = true;
            });

            if (maxLevel !== -1 && current >= maxLevel) {
                return;
            }
			
            children.forEach(child => {
                const nextChildren = BodyPartsData.containedChildren[child];
                if (nextChildren) {
                    traverse(nextChildren, current + 1);
                }
            });
        };

        traverse(immediateChildren, 1);

        return BodyPartsUtils.transformMap(map);
    }

    /**
	 * Returns all the contained ancestors.
	 */
    public getAllContainedAncestors (): BodyPart[] {
        let last = BodyPartsData.containedAncestors[this.data.radlexId];
        const map: BodyPartsMap = {};

        while (last && !map[last]) {
            map[last] = true;
            last = BodyPartsData.containedAncestors[last];
        }

        return BodyPartsUtils.transformMap(map);
    }

    /**
	 * Returns if the bodyPart is a descendant of a parent.
	 * @param {string} parentRadlexId - The parent radlex id.
	 */
    public isContained (parentRadlexId: string): boolean {
        const allAncestors = this.getAllContainedAncestors();
        
        for (let i = 0; i < allAncestors.length; i++) {
            if (allAncestors[i].getData().radlexId === parentRadlexId) {
                return true;
            }
        }

        return false;
    }

}

export {
    BodyPart
};