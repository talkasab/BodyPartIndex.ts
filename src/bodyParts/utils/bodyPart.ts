import { BodyPart } from '../bodyPart/bodyPart';
import { BODY_PARTS } from '..';

export type BodyPartsMap = { [key: string]: boolean };

/**
 * Transforms a map into an array of BodyParts.
 * @param {BodyPartsMap} map - The map.
 */
export const transformMapToBodyParts = (map: BodyPartsMap): BodyPart[] => {
    const arr = [];
	
    for (const i in map) {
        const data = BODY_PARTS.map[i];
        if (data && map[i]) {
            const bodyPart = new BodyPart(data);
            arr.push(bodyPart);
        }
    }

    return arr;
};

/**
 * Returns a BodyPart by radlex id.
 * @param {string | null} id - The id.
 */
export const getBodyPartById = (id: string | null): BodyPart | null => {
    if (!id) {
        return null;
    }

    const data = BODY_PARTS.map[id];

    if (!data) {
        return null;
    }

    return new BodyPart(data);
};