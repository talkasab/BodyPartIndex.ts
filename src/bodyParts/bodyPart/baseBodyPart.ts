import { IBodyPart } from 'src/bodyParts/interfaces/IBodyPart';
import { getBodyPartById } from 'src/bodyParts/utils/bodyPart';
import { BodyPart } from 'src/bodyParts/bodyPart/bodyPart';

export abstract class BaseBodyPart {

    /**
	 * Initializes the BodyPart.
	 * @param {IBodyPart} data - The data.
	 */
    public constructor (
		protected data: IBodyPart
    ) { }
	
    /**
	 * Returns the body part data.
	 */
    public getData (): IBodyPart {
        return this.data;
    }

    /**
	 * Returns the containedById.
	 */
    public getContainedById (): string {
        return this.data.containedById;
    }

    /**
	 * Returns the containedBy BodyPart.
	 */
    public getContainedBy (): BodyPart {
        return getBodyPartById(this.getContainedById()) as BodyPart;
    }

    /**
	 * Returns the containedById.
	 */
    public getPartOfId (): string | null {
        return this.data.partOfId || null;
    }

    /**
	 * Returns the partOf BodyPart.
	 */
    public getPartOf (): BodyPart | null {
        return getBodyPartById(this.getPartOfId());
    }

    /**
	 * Returns the leftId.
	 */
    public getLeftId (): string | null {
        return this.data.leftId || null;
    }

    /**
	 * Returns the left BodyPart.
	 */
    public getLeft (): BodyPart | null {
        return getBodyPartById(this.getLeftId());
    }

    /**
	 * Returns the rightId.
	 */
    public getRightId (): string | null {
        return this.data.rightId || null;
    }
	
    /**
	 * Returns the right BodyPart.
	 */
    public getRight (): BodyPart | null {
        return getBodyPartById(this.getRightId());
    }

    /**
	 * Returns the unsidedId.
	 */
	 public getUnsidedId (): string | null {
        return this.data.unsidedId || null;
    }
	
    /**
	 * Returns the unsided BodyPart.
	 */
    public getUnsided (): BodyPart | null {
        return getBodyPartById(this.getUnsidedId());
    }
		
}