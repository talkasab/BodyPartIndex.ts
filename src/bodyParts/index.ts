import { IBodyPartsResponse } from 'src/bodyParts/interfaces/IBodyPartsResponse';
import { BodyPart } from 'src/bodyParts/bodyPart/bodyPart';
import { getBodyParts } from 'src/bodyParts/utils/data';
import { configuration } from 'src/configuration';

export let BODY_PARTS: IBodyPartsResponse;

export class BodyPartIndex {

    /**
	 * Initializes the BodyPartIndex.
	 */
    public constructor () {
        BODY_PARTS = getBodyParts(configuration.file);
    }

    /**
	 * Returns a BodyPart object by radlexId.
	 * @param {string} radlexId - The radlexId.
	 */
    public getById (radlexId: string): BodyPart | null {
        const data = BODY_PARTS.bodyPartsMap[radlexId];

        if (!data) {
            return null;
        }

        return new BodyPart(data);
    }

    /**
	 * Returns the BodyParts that have a description or synonyms corresponding to a search value.
	 * @param {string} searchValue - The search value.
	 */
    public search (searchValue: string): BodyPart[] {
        const bodyParts = [];

        if (searchValue.length < 4) {
            return [];
        }
        
        for (const key in BODY_PARTS.bodyPartsMap) {
            const data = BODY_PARTS.bodyPartsMap[key];
            const bodyPartTerms = [
                data.description,
                ...data.synonyms || []
            ].join(' ');

            if (bodyPartTerms.indexOf(searchValue) !== -1) {
                const bodyPart = new BodyPart(data);
                bodyParts.push(bodyPart);
            }
        }

        return bodyParts;
    }

}