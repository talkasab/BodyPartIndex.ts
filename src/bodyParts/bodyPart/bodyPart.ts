import { ChildrenMap, AncestorsMap } from 'src/bodyParts/interfaces/IBodyPartsResponse';
import { BodyPartsMap, transformMapToBodyParts } from 'src/bodyParts/utils/bodyPart';
import { BaseBodyPart } from 'src/bodyParts/bodyPart/baseBodyPart';
import { BODY_PARTS } from 'src/bodyParts';

export class BodyPart extends BaseBodyPart {

    /**
	 * Returns the immediate contained children.
	 */
    public getImmediateContainedChildren (): BodyPart[] {
        return this.getChildren(BODY_PARTS.containedChildren, 1);
    }

    /**
	 * Returns all the contained children.
	 */
    public getAllContainedChildren (): BodyPart[] {
        return this.getChildren(BODY_PARTS.containedChildren);
    }

    /**
	 * Returns all the contained ancestors.
	 */
    public getAllContainedAncestors (): BodyPart[] {
        return this.getAllAncestors(BODY_PARTS.containedAncestors);
    }

    /**
	 * Returns if the bodyPart is a contained descendant of a parent.
	 * @param {string | BodyPart} parent - The parent.
	 */
    public isContained (parent: string | BodyPart): boolean {
        return this.isDescendant(parent, this.getAllContainedAncestors());
    }

    /**
	 * Returns the immediate partOf children.
	 */
    public getImmediatePartOfChildren (): BodyPart[] {
        return this.getChildren(BODY_PARTS.partOfChildren, 1);
    }

    /**
	 * Returns all the partOf children.
	 */
    public getAllPartOfChildren (): BodyPart[] {
        return this.getChildren(BODY_PARTS.partOfChildren);
    }

    /**
	 * Returns all the partOf ancestors.
	 */
    public getAllPartOfAncestors (): BodyPart[] {
        return this.getAllAncestors(BODY_PARTS.partOfAncestors);
    }

    /**
	 * Returns if the BodyPart is a partOf descendant of a parent.
	 * @param {string | BodyPart} parent - The parent.
	 */
    public isPartOf (parent: string | BodyPart): boolean {
        return this.isDescendant(parent, this.getAllPartOfAncestors());
    }

    /**
	 * Returns the children.
	 * @param {ChildrenMap} children - The children.
	 * @param {number} depth - The depth (if specified will return children up to that level).
	 */
    private getChildren (children: ChildrenMap, depth = -1): BodyPart[] {
        const immediateChildren = children[this.data.radlexId];
        const map: BodyPartsMap = {};
		
        if (!immediateChildren) {
            return [];
        }

        const traverse = (row: string[], current: number): void => {
            row.forEach(child => {
                map[child] = true;
            });

            if (depth !== -1 && current >= depth) {
                return;
            }
			
            row.forEach(child => {
                const nextChildren = children[child];
                if (nextChildren) {
                    traverse(nextChildren, current + 1);
                }
            });
        };

        traverse(immediateChildren, 1);

        return transformMapToBodyParts(map);
    }

    /**
	 * Returns all the ancestors.
	 * @param {AncestorsMap} ancestors - The ancestors.
	 */
    private getAllAncestors (ancestors: AncestorsMap): BodyPart[] {
        let last = ancestors[this.data.radlexId];
        const map: BodyPartsMap = {};

        while (last && !map[last]) {
            map[last] = true;
            last = ancestors[last];
        }

        if (map[this.data.radlexId]) {
            delete map[this.data.radlexId];
        }

        return transformMapToBodyParts(map);
    }

    /**
	 * Returns if the BodyPart is a descendant of a parent.
	 * @param {string | BodyPart} parent - The parent.
	 * @param {BodyPart[]} ancestors - The ancestors.
	 */
    private isDescendant (parent: string | BodyPart, ancestors: BodyPart[]): boolean {
        const parentRadlexId = parent instanceof BodyPart ? 
            parent.getData().radlexId : parent;

        for (let i = 0; i < ancestors.length; i++) {
            if (ancestors[i].getData().radlexId === parentRadlexId) {
                return true;
            }
        }

        return false;
    }

}