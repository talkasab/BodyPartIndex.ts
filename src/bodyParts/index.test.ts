import { BodyPartIndex, BODY_PARTS } from 'src/bodyParts/index';
import { BodyPart } from 'src/bodyParts/bodyPart/bodyPart';
import { configuration } from 'src/configuration';

const bodyPart1 = {
    codes: [ { code: '78320081', system: 'SNOMED' } ],
    containedById: 'RID901',
    description: 'hand',
    partOfId: 'RID902',
    radlexId: 'RID901'
};

const bodyPart2 = {
    codes: [ { code: '78320082', system: 'SNOMED' } ],
    containedById: 'RID901',
    description: 'feet',
    radlexId: 'RID902'
};

const bodyPart3 = {
    codes: [ { code: '78320083', system: 'SNOMED' } ],
    containedById: 'RID901',
    description: 'leg',
    radlexId: 'RID903',
    synonyms: [
        'feet'
    ]
};

const bodyPart4 = {
    containedById: 'RID903',
    description: 'eye',
    partOfId: 'RID902',
    radlexId: 'RID904',
    synonyms: [
        'aaa',
        'ccc'
    ]
};

const bodyPart5 = {
    codes: [ { code: '78320085', system: 'SNOMED' } ],
    containedById: 'RID904',
    description: 'finger',
    partOfId: 'RID901',
    radlexId: 'RID905'
};

const fileMock = {
    $schema: 'MY_SCHEMA',
    $version: '2.0.0',
    bodyParts: [ bodyPart1, bodyPart2, bodyPart3, bodyPart4, bodyPart5 ]
};

describe('setUpConfiguration()', () => {

    beforeEach(() => {
        configuration.file = fileMock;
        new BodyPartIndex({
            localBodyPartMappings: [
                {
                    localCode: {
                        code: 'AAABBBCC',
                        system: 'LOCAL'
                    },
                    radlexId: 'RID901'
                }
            ]
        });
    });

    it ('matches the BODY_PARTS object', () => {
        expect(BODY_PARTS).toMatchSnapshot();
    });

});

describe('get()', () => {
    let index: BodyPartIndex;

    beforeEach(() => {
        configuration.file = fileMock;
        index = new BodyPartIndex({
            localBodyPartMappings: [
                {
                    localCode: {
                        code: 'AAABBBCC',
                        system: 'LOCAL'
                    },
                    radlexId: 'RID901'
                },
                {
                    localCode: {
                        code: 'EEEFFFGGGHHH',
                        system: 'LOCAL'
                    },
                    // Non-existing BodyPart
                    radlexId: 'RID901222'
                }
            ]
        });
    });

    it ('returns null for non-existing radlexId', () => {
        expect(index.get('RID9052')).toEqual(null);
    });

    it ('returns the correct BodyPart by radlexId', () => {
        expect(index.get('RID905')).toEqual(new BodyPart(bodyPart5));
    });

    it ('returns the correct BodyPart by code', () => {
        expect(index.get('78320083')).toEqual(new BodyPart(bodyPart3));
    });

    it ('returns the correct BodyPart by local code', () => {
        expect(index.get('AAABBBCC')).toEqual(
            new BodyPart({ 
                ...bodyPart1, 
                codes: [
                    ...bodyPart1.codes,
                    {
                        code: 'AAABBBCC',
                        system: 'LOCAL'
                    }
                ] 
            })
        );
    });

});

describe('search()', () => {
    let index: BodyPartIndex;

    beforeEach(() => {
        configuration.file = fileMock;
        index = new BodyPartIndex();
    });

    it ('returns empty array for searchValue less than 3', () => {
        expect(index.search('av')).toEqual([]);
    });

    it ('returns the correct BodyPart (eye)', () => {
        expect(index.search('eye')).toEqual([ new BodyPart(bodyPart4) ]);
    });

    it ('returns the correct BodyPart (eye)', () => {
        expect(index.search('aaa')).toEqual([ new BodyPart(bodyPart4) ]);
    });

    it ('returns the correct BodyParts (feet)', () => {
        expect(index.search('feet')).toEqual([ 
            new BodyPart(bodyPart2), 
            new BodyPart(bodyPart3) 
        ]);
    });

    it ('returns empty array when no BodyParts are found', () => {
        expect(index.search('feet2')).toEqual([]);
    });
});