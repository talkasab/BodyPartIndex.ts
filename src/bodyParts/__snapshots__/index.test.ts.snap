// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`setUpConfiguration() matches the BODY_PARTS object 1`] = `
{
  "codes": {
    "78320081": [
      "RID901",
    ],
    "78320082": [
      "RID902",
    ],
    "78320083": [
      "RID903",
    ],
    "78320085": [
      "RID905",
    ],
    "AAABBBCC": [
      "RID901",
    ],
  },
  "containedAncestors": {
    "RID901": "RID901",
    "RID902": "RID901",
    "RID903": "RID901",
    "RID904": "RID903",
    "RID905": "RID904",
  },
  "containedChildren": {
    "RID901": [
      "RID902",
      "RID903",
    ],
    "RID903": [
      "RID904",
    ],
    "RID904": [
      "RID905",
    ],
  },
  "map": {
    "RID901": {
      "codes": [
        {
          "code": "78320081",
          "system": "SNOMED",
        },
        {
          "code": "AAABBBCC",
          "system": "LOCAL",
        },
      ],
      "containedById": "RID901",
      "description": "hand",
      "partOfId": "RID902",
      "radlexId": "RID901",
    },
    "RID902": {
      "codes": [
        {
          "code": "78320082",
          "system": "SNOMED",
        },
      ],
      "containedById": "RID901",
      "description": "feet",
      "radlexId": "RID902",
    },
    "RID903": {
      "codes": [
        {
          "code": "78320083",
          "system": "SNOMED",
        },
      ],
      "containedById": "RID901",
      "description": "leg",
      "radlexId": "RID903",
      "synonyms": [
        "feet",
      ],
    },
    "RID904": {
      "containedById": "RID903",
      "description": "eye",
      "partOfId": "RID902",
      "radlexId": "RID904",
      "synonyms": [
        "aaa",
        "ccc",
      ],
    },
    "RID905": {
      "codes": [
        {
          "code": "78320085",
          "system": "SNOMED",
        },
      ],
      "containedById": "RID904",
      "description": "finger",
      "partOfId": "RID901",
      "radlexId": "RID905",
    },
  },
  "partOfAncestors": {
    "RID901": "RID902",
    "RID904": "RID902",
    "RID905": "RID901",
  },
  "partOfChildren": {
    "RID901": [
      "RID905",
    ],
    "RID902": [
      "RID901",
      "RID904",
    ],
  },
  "version": "2.0.0",
}
`;

exports[`setUpConfiguration() with local JSON file matches the BODY_PARTS object 1`] = `
{
  "codes": {
    "78320081": [
      "RID901",
    ],
    "78320082": [
      "RID902",
    ],
    "78320083": [
      "RID903",
    ],
    "78320085": [
      "RID905",
    ],
    "AAABBBCC": [
      "RID901",
    ],
  },
  "containedAncestors": {
    "RID901": "RID901",
    "RID902": "RID901",
    "RID903": "RID901",
    "RID904": "RID903",
    "RID905": "RID904",
  },
  "containedChildren": {
    "RID901": [
      "RID902",
      "RID903",
    ],
    "RID903": [
      "RID904",
    ],
    "RID904": [
      "RID905",
    ],
  },
  "map": {
    "RID901": {
      "codes": [
        {
          "code": "78320081",
          "system": "SNOMED",
        },
        {
          "code": "AAABBBCC",
          "system": "LOCAL",
        },
      ],
      "containedById": "RID901",
      "description": "hand",
      "partOfId": "RID902",
      "radlexId": "RID901",
    },
    "RID902": {
      "codes": [
        {
          "code": "78320082",
          "system": "SNOMED",
        },
      ],
      "containedById": "RID901",
      "description": "feet",
      "radlexId": "RID902",
    },
    "RID903": {
      "codes": [
        {
          "code": "78320083",
          "system": "SNOMED",
        },
      ],
      "containedById": "RID901",
      "description": "leg",
      "radlexId": "RID903",
      "synonyms": [
        "feet",
      ],
    },
    "RID904": {
      "containedById": "RID903",
      "description": "eye",
      "partOfId": "RID902",
      "radlexId": "RID904",
      "synonyms": [
        "aaa",
        "ccc",
      ],
    },
    "RID905": {
      "codes": [
        {
          "code": "78320085",
          "system": "SNOMED",
        },
      ],
      "containedById": "RID904",
      "description": "finger",
      "partOfId": "RID901",
      "radlexId": "RID905",
    },
  },
  "partOfAncestors": {
    "RID901": "RID902",
    "RID904": "RID902",
    "RID905": "RID901",
  },
  "partOfChildren": {
    "RID901": [
      "RID905",
    ],
    "RID902": [
      "RID901",
      "RID904",
    ],
  },
  "version": "2.0.0",
}
`;
